include/envoy/network/filter.h

```diff
@@ -232,6 +234,47 @@ class ListenerFilterManager {
  */
 typedef std::function<void(ListenerFilterManager& filter_manager)> ListenerFilterFactoryCb;
 
+/**
+ * Interface representing a single filter chain.
+ */
+class FilterChain {
+public:
+  virtual ~FilterChain() {}
+
+  /**
+   * @return bool true if filter chain implements secure transport, false if it doesn't.
+   */
+  virtual bool implementsSecureTransport() const PURE;
+
+  /**
+   * @return Network::TransportSocketPtr a transport socket to be used by the new connection.
+   */
+  virtual TransportSocketPtr createTransportSocket() const PURE;
+
+  /**
+   * const std::vector<FilterFactoryCb>& a list of filters to be used by the new connection.
+   */
+  virtual const std::vector<FilterFactoryCb>& getNetworkFilterFactories() const PURE;
+};
+
+typedef std::shared_ptr<FilterChain> FilterChainSharedPtr;
+
+/**
+ * Interface for searching through configured filter chains.
+ */
+class FilterChainManager {
+public:
+  virtual ~FilterChainManager() {}
+
+  /**
+   * Find filter chain that's matching metadata from the new connection.
+   * @param socket supplies connection metadata that's going to be used for the filter chain lookup.
+   * @return const FilterChainSharedPtr filter chain to be used by the new connection,
+   *         nullptr if no matching filter chain was found.
+   */
+  virtual const FilterChainSharedPtr findFilterChain(ConnectionSocketPtr& socket) const PURE;
```

Comment:
nit: Could this be `ConnectionSocketConstPtr` ?

Also, is there a reason to return a shared pointer here? From my read of the code we never actually persistently store the result. Could we just return a `const FilterChain*` and avoid the shared_ptr inc/dec overhead?

Reply:
wrt `ConnectionSocketConstPtr` -  do you mean `const unique_ptr<ConnectionSocket>` (i.e. `const ConnectionSocketPtr`) or `unique_ptr<const ConnectionSocket>`?

wrt `FilterChainSharedPtr` - there is no reason why we couldn't return `const FilterChain*`, but I thought the whole idea behind `unique_ptr` / `shared_ptr` is to avoid passing raw pointers around.

Reply:
I meant `unique_ptr<const ConnectionSocket>`

For `FilterChainSharedPtr` raw pointer is fine in a case like this where we are just referencing constant data that might be nullptr.

Reply:
Erm, but we cannot safely create `ConnectionSocketConstPtr` from `ConnectionSocketPtr`, without move... and few lines below, we need `ConnectionSocketPtr` to pass into `createServerConnection` to create `ConnectionImpl`, which means that we would need to move twice.

Per @lizan's suggestion, perhaps we should pass `const ConnectionSocket*` here instead?